File: Bronze/2016/February/02_circular_barn/circular_barn2.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <climits>

using namespace std;

int main() {
    ifstream fin("cbarn.in");
    ofstream fout("cbarn.out");

    int N;
    fin >> N;

    vector<int> rooms(N);

    for (int i = 0; i < N; i++) {
        fin >> rooms[i];
    }

    int result = INT_MAX;

    // 以每个房间为起点，计算总距离
    for (int i = 0; i < N; i++) {
        int acc = 0;
        for (int j = 0; j < N; j++) {
            acc += rooms[(i + j) % N] * j;
        }
        result = min(result, acc);
    }

    fout << result << endl;
    fout.close();

    return 0;
}

File: Bronze/2016/February/02_circular_barn/CircularBarn2.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.PrintWriter;

/**
 * @author Red
 * @version 1.0
 */
public class CircularBarn2 {
    static int N;
    static int[] rooms;
    public static void main(String[] args) throws Exception {
        BufferedReader r = new BufferedReader(new FileReader("cbarn.in"));
        PrintWriter pw = new PrintWriter("cbarn.out");
        N = Integer.parseInt(r.readLine());
        // 记录每个房间牛的数量
        rooms = new int[N];

        for (int i = 0; i < N; i++) {
            rooms[i] = Integer.parseInt(r.readLine());
        }

        int result = Integer.MAX_VALUE;
        // 以每个房间为起点，计算总距离
        for (int i = 0; i < N; i++) {
            int acc = 0;
            for (int j = 0; j < N; j++) {
                acc += rooms[(i + j) % N] * j;
            }
            result = Math.min(result, acc);
        }

        pw.println(result);
        pw.close();
    }
}

